Perform validation later in PID pid binding script

ross-spencer opened the pull request:
When validation was added in 97b1ad0c4faa328fd2e9e34724d60743c029c9af it didn't take into account that all 'required' parameters might not be available at the time validation is done. We are interested in the fields of the 'handle' configuration form in Archivematica's dashboard. If some part of that form is incomplete at the time it arrives at validation in this PR, then we can exit early and provide some useful feedback to the user. We can ignore the elements expected to be added to the configuration [here](https://github.com/artefactual/archivematica/blob/85074e77db9dc9d10f44f0d5461d269b739fea6c/src/MCPClient/lib/clientScripts/bind_pids.py#L202) in the same script for example. 

Connected to archivematica/issues#776

ross-spencer committed:
Separate validation responsibilities

We want to validate the handle server configuration early, but only
on the values we know that we will have. This commit sees a change to
the Bind PIDs code that pulls apart fields required to bind and
attach PIDs in Archivematica but calls their validation methods at
more appropriate points in the execution of the code.

replaceafill commented on src/MCPClient/lib/clientScripts/bind_pids.py:
```diff
@@ -224,7 +224,9 @@ def main(job, sip_uuid, shared_path, bind_pids_switch):
     try:
-        _validate(handle_config)
+        # We know entity_type and desired_pid isn't provided until later, we
+        # can't validate without ignoring it here.
```
This comment is a bit confusing if you're not familiar with the code. It's explaining what is about to happen not why (Why are those parameters set in `_bind_pid_to_model`?)

ross-spencer replied:
Good point. I've decided to rework this and will re-submit for re-review.

replaceafill commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -206,12 +206,17 @@ class BindPIDException(Exception):
 
 
-def _validate(argdict):
+def _validate(argdict, ignore_values=()):
```
Could we explain the `ignore_values` parameter in the docstring?

replaceafill commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -206,12 +206,17 @@ class BindPIDException(Exception):
         val = argdict.get(param)
         if not val:
             raise BindPIDException("A value for parameter {} is required".format(param))
+    # In cases we haven't the entity type we are looking for, return early.
```
This sounds like we're "getting" the value of entity type and we don't like what we get so we return. Like what it's happening a couple of lines below (`et_validation = ENTITY_TYPES.get(entity_type)`). We're returning early because the caller decided we could do so, correct?

ross-spencer replied:
It does. It has made me think twice about this approach.

replaceafill reviewed (approved):
@ross-spencer The change works but is a bit hard to understand. The early return added to `bindpid._validate` didn't help either.

It made me wonder if it'd be better to remove `entity_type` and `desired_pid` from `REQ_PARAMS`. It feels like "they're not really required anymore under **some** circunstances". And I don't see any negative implications to `bind_pid._get_bind_pid_config` which seems to be the only other caller of `bindpid._validate` and which makes sure both parameters are set before the call.

ross-spencer commented:
Hi @replaceafill I've reworked the PR into something a little more sensible. Your questions were insightful. I held back a little in the original PR because I didn't want to pull apart the code too much, but I do think the decisions should still be sound. Holding back still and focusing on `bind_pids` I've separated validation responsibilities and so I was able to remove the `ignore=` and then also make sure validation is happening exactly where it should be. I wrapped my two new functions in a new function with the `_validate` function signature so that it doesn't break `bind_pids`. I feel like there is still a little more we might be able to do for these scripts (they're not being introduced in `1.10` so perhaps any more broad changes can be pushed into a later release depending on how significant they look/feel.

Anyhoo!! It's ready to re-review for now!! 

Related, I don't know if you keep your own `Makefile` for am.git, (you might already have made the following modification) but I've added a recipe PR here for testing `archivematica-common`. As this change touches on a module in that component of Archivematica you can test that I haven't broken anything using this. 

* https://github.com/artefactual-labs/am/pull/84

replaceafill commented:
Thank you @ross-spencer Sorry for putting extra work on you but this is easier to grasp in my opinion. :+1: from me.

My `Makefile` is a bunch of customizations :blush: So far I've been testing with tox directly in a separate virtualenv (similarly to how travis does it).

ross-spencer commented:
@replaceafill worth it I think too! :) 

RE: `Makefile` we need to have a dev day and share what we're all doing - I feel there's a lot to learn from each other :nerd_face:
